From d1f9b85bf6de868345a9f00d2f39aba683d1a49b Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sat, 8 Nov 2008 21:05:23 +0000
Subject: [PATCH] CONFIGURATION-345: Ensure that "ISO-8859-1" is used as
 default encoding, even by the constructors of the super class.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@712431 13f79535-47bb-0310-9956-ffa450edef68
---
 .../PropertiesConfiguration.java              | 18 ++++++++---
 .../TestPropertiesConfiguration.java          | 31 +++++++++++++++++++
 xdocs/changes.xml                             |  4 +++
 3 files changed, 48 insertions(+), 5 deletions(-)

diff --git a/src/java/org/apache/commons/configuration/PropertiesConfiguration.java b/src/java/org/apache/commons/configuration/PropertiesConfiguration.java
index 96dfa43f55..8924c0585b 100644
--- a/src/java/org/apache/commons/configuration/PropertiesConfiguration.java
+++ b/src/java/org/apache/commons/configuration/PropertiesConfiguration.java
@@ -213,11 +213,6 @@ public class PropertiesConfiguration extends AbstractFileConfiguration
     /** Allow file inclusion or not */
     private boolean includesAllowed;
 
-    // initialization block to set the encoding before loading the file in the constructors
-    {
-        setEncoding(DEFAULT_ENCODING);
-    }
-
     /**
      * Creates an empty PropertyConfiguration object which can be
      * used to synthesize a new Properties file by adding values and
@@ -343,6 +338,19 @@ public void setHeader(String header)
         getLayout().setHeaderComment(header);
     }
 
+    /**
+     * Returns the encoding to be used when loading or storing configuration
+     * data. This implementation ensures that the default encoding will be used
+     * if none has been set explicitly.
+     *
+     * @return the encoding
+     */
+    public String getEncoding()
+    {
+        String enc = super.getEncoding();
+        return (enc != null) ? enc : DEFAULT_ENCODING;
+    }
+
     /**
      * Returns the associated layout object.
      *
diff --git a/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java b/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java
index ddc6b222f8..002d2dab81 100644
--- a/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java
@@ -21,6 +21,7 @@
 import java.io.FileOutputStream;
 import java.io.FileWriter;
 import java.io.IOException;
+import java.io.InputStream;
 import java.io.OutputStream;
 import java.io.PrintWriter;
 import java.io.Reader;
@@ -825,6 +826,36 @@ public void testSaveWithDataConfig() throws ConfigurationException
         assertEquals("Property not saved", "bar", config2.getString("foo"));
     }
 
+    /**
+     * Tests whether the correct default encoding is used when loading a
+     * properties file. This test is related to CONFIGURATION-345.
+     */
+    public void testLoadWithDefaultEncoding() throws ConfigurationException
+    {
+        class PropertiesConfigurationTestImpl extends PropertiesConfiguration
+        {
+            String loadEncoding;
+
+            public PropertiesConfigurationTestImpl(String fileName)
+                    throws ConfigurationException
+            {
+                super(fileName);
+            }
+
+            public void load(InputStream in, String encoding)
+                    throws ConfigurationException
+            {
+                loadEncoding = encoding;
+                super.load(in, encoding);
+            }
+        }
+
+        PropertiesConfigurationTestImpl testConf = new PropertiesConfigurationTestImpl(
+                testProperties);
+        assertEquals("Default encoding not used", "ISO-8859-1",
+                testConf.loadEncoding);
+    }
+
     /**
      * Creates a configuration that can be used for testing copy operations.
      *
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 93a0266c77..3db6dc34db 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -38,6 +38,10 @@
         configuration nodes for properties with multiple values. This
         improves compatibility with queries.
       </action>
+      <action dev="oheger" type="fix" issue="CONFIGURATION-345">
+        PropertiesConfiguration now per default uses the encoding "ISO-8859-1"
+        for loading properties files.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-344">
         CombinedConfiguration could cause a deadlock when it was accessed while
         concurrently a reload of one of its child configuration happened. This
